### PR TITLE
Improve trace logging in TransportReplicationAction and error reporting at RecoveryWhileUnderLoadIT

### PR DESCRIPTION
**Pull request by bleskes**
Things that helped me traced down an issue.

### CODE REVIEW

File: core/src/test/java/org/elasticsearch/recovery/RecoveryWhileUnderLoadIT.java [hunk 1]
```diff
@@ -255,19 +256,21 @@ public void testRecoverWhileRelocating() throws Exception {
             logger.info("--> refreshing the index");
             refreshAndAssert();
             logger.info("--> verifying indexed content");
-            iterateAssertCount(numShards, indexer.totalIndexedDocs(), 10);
+            iterateAssertCount(numShards, (int) indexer.totalIndexedDocs(), 10);
```
Comment (jasontedor) [inline core/src/test/java/org/elasticsearch/recovery/RecoveryWhileUnderLoadIT.java]: I'm missing the point of this casts; the corresponding method parameter is a `long`?

File: core/src/test/java/org/elasticsearch/recovery/RecoveryWhileUnderLoadIT.java [hunk 2]
```diff
@@ -255,19 +256,21 @@ public void testRecoverWhileRelocating() throws Exception {
...
     }
 
     private void iterateAssertCount(final int numberOfShards, final long numberOfDocs, final int iterations) throws Exception {
         SearchResponse[] iterationResults = new SearchResponse[iterations];
         boolean error = false;
+        SearchResponse lastErroneousResponse = null;
         for (int i = 0; i < iterations; i++) {
-            SearchResponse searchResponse = client().prepareSearch().setSize(0).setQuery(matchAllQuery()).get();
+            SearchResponse searchResponse = client().prepareSearch().setSize((int) numberOfDocs).setQuery(matchAllQuery()).addSort("id", SortOrder.ASC).get();
             logSearchResponse(numberOfShards, numberOfDocs, i, searchResponse);
```
Comment (jasontedor) [inline core/src/test/java/org/elasticsearch/recovery/RecoveryWhileUnderLoadIT.java]: I'm missing the point of this casts; the corresponding method parameter is a `long`?

File: core/src/test/java/org/elasticsearch/recovery/RecoveryWhileUnderLoadIT.java [hunk 3]
```diff
@@ -255,19 +256,21 @@ public void testRecoverWhileRelocating() throws Exception {
...
         }
     }
 
     private void iterateAssertCount(final int numberOfShards, final long numberOfDocs, final int iterations) throws Exception {
         SearchResponse[] iterationResults = new SearchResponse[iterations];
         boolean error = false;
+        SearchResponse lastErroneousResponse = null;
         for (int i = 0; i < iterations; i++) {
-            SearchResponse searchResponse = client().prepareSearch().setSize(0).setQuery(matchAllQuery()).get();
+            SearchResponse searchResponse = client().prepareSearch().setSize((int) numberOfDocs).setQuery(matchAllQuery()).addSort("id", SortOrder.ASC).get();
```
Comment (jasontedor) [inline core/src/test/java/org/elasticsearch/recovery/RecoveryWhileUnderLoadIT.java]: I'm missing the point of this casts; the corresponding method parameter is a `long`?

**Comment by jasontedor** [comment]
I think that all of the calls to `logger#trace` touched in this change set should be wrapped in `if`-blocks guarded by the condition `logger.isTraceEnabled()` lest we always be needlessly allocating object arrays (because of the varargs parameter).

File: core/src/test/java/org/elasticsearch/recovery/RecoveryWhileUnderLoadIT.java (hunk 1 above)
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/recovery/RecoveryWhileUnderLoadIT.java]: left over - I tried to move away from the long param, but realized it caused more pain that gains. reverting...

File: core/src/test/java/org/elasticsearch/recovery/RecoveryWhileUnderLoadIT.java (hunk 3 above)
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/recovery/RecoveryWhileUnderLoadIT.java]: the size of a search is an int, which is good...

Comment (jasontedor) [inline core/src/test/java/org/elasticsearch/recovery/RecoveryWhileUnderLoadIT.java]: Oops, I thought I deleted this comment after I realized it was made in error. Sorry. :)

**Comment by bleskes** [comment]
@jasontedor thanks. I pushed an update. I chose to only protect the trace logs which are done under "normal" operation. The exceptional cases are not worth the extra if imho.

**Comment by jasontedor** [comment]
@bleskes Okay. LGTM.

**Comment by imotov** [comment]
testRecoverWhileRelocating failed today:
https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+master+multijob-unix-compatibility/os=debian/112/console

**Pull request closed without merging**